activerecord/lib/active_record/timestamp.rb

```diff
@@ -59,19 +59,26 @@ def touch_attributes_with_time(*names, time: nil)
         attribute_names.index_with(time || current_time_from_proper_timezone)
       end
 
-      private
-        def timestamp_attributes_for_create_in_model
-          timestamp_attributes_for_create.select { |c| column_names.include?(c) }
-        end
+      def timestamp_attributes_for_create_in_model
+        @timestamp_attributes_for_create_in_model ||=
+          (timestamp_attributes_for_create & column_names).freeze
+      end
 
-        def timestamp_attributes_for_update_in_model
-          timestamp_attributes_for_update.select { |c| column_names.include?(c) }
-        end
+      def timestamp_attributes_for_update_in_model
+        @timestamp_attributes_for_update_in_model ||=
+          (timestamp_attributes_for_update & column_names).freeze
+      end
 
-        def all_timestamp_attributes_in_model
-          timestamp_attributes_for_create_in_model + timestamp_attributes_for_update_in_model
-        end
+      def all_timestamp_attributes_in_model
+        @all_timestamp_attributes_in_model ||=
+          (timestamp_attributes_for_create_in_model + timestamp_attributes_for_update_in_model).freeze
+      end
 
+      def current_time_from_proper_timezone
+        default_timezone == :utc ? Time.now.utc : Time.now
+      end
+
+      private
         def timestamp_attributes_for_create
           ["created_at", "created_on"]
```

Comment:
Should we extract these to a constant that we return here?

Reply:
I leaved these as it is, since the extraction has no performance gain.

As the result of allocation tracer shows, as long as memoizing the result of `timestamp_attributes_for_{create,update}_in_model`, these internal methods (`timestamp_attributes_for_{create,update}`) will only be called once per model.

```
% git grep -n '\btimestamp_attributes_for_create\b'
activerecord/lib/active_record/timestamp.rb:64:          (timestamp_attributes_for_create & column_names).freeze
activerecord/lib/active_record/timestamp.rb:82:        def timestamp_attributes_for_create
% git grep -n '\btimestamp_attributes_for_update\b'
activerecord/lib/active_record/timestamp.rb:69:          (timestamp_attributes_for_update & column_names).freeze
activerecord/lib/active_record/timestamp.rb:86:        def timestamp_attributes_for_update
```